Update contributing guidelines (#311)
ManimCommunity/manim: CONTRIBUTING.md

# Contributing

Thank you for contributing to Manim!  However you have decided to contribute or
interact with the community, please always be civil and respect other members
of the community.  Manim is a free open source software for mathematical
animations, and as such we welcome **everyone** who is interested in
mathematics, pedagogy, computer animations, open source, software development,
and beyond.  Manim accepts contributions of many kinds, detailed below.

## Reporting bugs

One of the best ways of contributing to Manim is by reporting bugs.  If you
have encountered something that you believe is a bug, please follow these
steps:

1. Make sure to search for other users who may have had similar issues in the
   past.  Search the repository's issues page (don't forget to search closed
   issues), bring it up on Discord, use sites like StackOverflow, and exercise
   your best Google practices.  If you can't find anything helpful, then go to
   the next step.

> 1. Make sure you can reproduce the issue, i.e. that you have some code that

Reviewer: Issue : Indexation mistake

Author: You can actually use any numbers for lists in markdown, and they will be displayed correctly:

1. this list
1. has all indices
1. equal to one

Reviewer 2: i mean, sure, but as a md file this looks weird and inconsistent, like, either 1. 1. 1. 1. or 1. 2. 3. 4. ,  not 1. 1. 2. 3. 4. as that has no apparent pattern

Author: Ok, done!

Reviewer: > You can actually use any numbers for lists in markdown, and they will be displayed correctly:

TIL that :D

   illustrates the bug **every time** (or at least most of the time) it is
   executed.
   
2. Make sure to clarify what behavior you expected, and how the actual behavior
   differs from your expectation.

3. Make sure to gather information about your environment, such as your

>    operating system, manim version, python version, and any stack traces that

Reviewer 3: I think it's best to just say to make sure you're using the latest version; I don't think we should support multiple versions or anything like that.

Reviewer 4: I am here with eulertour

Author: Done

   the code may have generated (if applicable).
   
4. Please open an issue only after you have gathered this information.  A
   community member will (hopefully) respond and start a conversation to
   address the issue.
   
5. Please give the community a reasonable amount of time before asking again,
   or insisting on your issue.  Keep in mind that everyone is a volunteer.  If
   you wait for a reasonable amount of time and you receive no response, feel

>    free to insist.

Reviewer 4: Suggestion: "feel free to insist." 
to "feel free to ask again"
sounds maybe a little more gentle.

Author: Done

## Contributing code

Many ways of contributing will involve writing, reading, testing, or
refactoring code.  As our repository is a Fork of [Manim by
3b1b](https://github.com/3b1b/manim), contributing in this way can be a bit
confusing.  Here is a short guide on how to do it.

> 1. First, make a fork of this repository.  This creates your own copy of the

Reviewer 3: Do you think it'd be better to just link to a tutorial on how to do this? I get the feeling there must be some tutorial out there which is more comprehensive, and it'd save us the trouble of answering issues on people's git setups.

Author: How about this one? https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow. We can delete a good chunk of our document in favor of that one instead.

   whole codebase on GitHub.

2. Clone the repository you forked using the command below (bash/zsh).

   ```sh
     git clone <your-fork-url>
     cd manim
   ```

	Note: do not `git clone` the original ManimCommunity repository.  You must
    clone your own fork.  After this step, there are three different
    repositories to keep track of: the original ManimCommunity repo, your own
    fork of it, and your local repository.

3. You have to make `git` be aware of the three repositories, by entering the
   commands below.

   ```sh
   git remote add upstream https://github.com/ManimCommunity/manim.git

>    git pull upstream master

Reviewer 5: You missed the checkout command. I think the user need to create a new branch locally before actually pushing it.
```suggestion
   git pull upstream master
   git checkout -b <your-branch-name> upstream/master
```

Author: Ah good point. Done

>    git remote add upstream https://github.com/ManimCommunity/manim.git
>    git pull upstream master

Reviewer 6: ```suggestion
   git remote add community https://github.com/ManimCommunity/manim.git
   git pull community master
```

Nitpick: I think naming the remote `community` over `upstream` will provide more clarity as to what exactly the remote is.

Author: I disagree here. I think upstream is more standard (and it's also shorter).

   ```

	After these commands, your local repository can keep track of your fork
    (referred to as 'origin') as well as the main ManimCommunity repository

>     (referred to as 'upstream').

Reviewer 6: ```suggestion
    (referred to as 'community').
```

4. You can now make any changes to your local repository, add them, and commit
   them.

5. To make your changes visible in your GitHub fork, instead of typing in `git
   push` as usual, you need to enter the command below.

   ```sh
   git push -u origin <your-branch-name>
   ```

   Doing so creates a new branch with the updated contents of your fork on
   GitHub (the 'origin').

6. If you want your changes to be incorporated to the main ManimCommunity
   repository, you need to make sure that there are no merge conflicts between
   the current upstream/master and the changes you are trying to make.  For
   this purpose, please execute `git pull origin master` before you push
   changes to your fork.  If this generates merge conflicts, make sure to
   resolve them before submitting a PR.

7. Further, make sure that you run the tests locally and that they all pass
   before submitting a PR.  You can run the tests locally by executing `pytest`
   from the main `manim/` folder (do not execute pytest from the code folder

>    `manim/manim`).

Reviewer 4: Is this really necessary? 
I normally don't do this.
I thought that github tests are there for this exact reason, that you don't have to run them locally.
Especially for new contributers, I think it would be nice to keep it as simple as possible

Reviewer: Hmmm I disagree here. Asking people to test locally will avoid 300 commits trying to fix tests on GitHub action.

Author: > Is this really necessary? I normally don't do this.

You should!

> I thought that github tests are there for this exact reason, that you don't have to run them locally.

No. github tests run the tests on many platforms and many versions of different things. The author of a PR should be expected to run tests locally.

> Especially for new contributers, I think it would be nice to keep it as simple as possible

I assume you mean "inexperienced contributors" and not "new contributors". A new contributor to manim may already have a lot of experienced with open source and not be intimidate by local testing. Further, I don't think that asking people to run a single command with a single word (`pytest`) is asking too much.

8. Finally, make sure to use the `black` autoformatter before submitting a PR.
   You can execute it by running `black <your_file>`.  This will make changes

>    to your code, and you will need to add/commit those changes.

Reviewer 4: For me, this is a step that is a little annoying, primary because there is no good IDE support yet (I am using pycharm, maybe this is easier for other IDEs?) . Especially when you have more then one file, making this with every single file can get quit annoying. But maybe this is a point we can discuss in a separate issue. For now I am fine to leave it like this and get this pull request merged with this procedure.

Reviewer 7: You can just run `black .` in the manim directory and it'll autoformat all files there, if they need to be changed; and given that the whole library has been preformatted, only the files you've changed will get formatted.

Reviewer: Agree with @kilacoda . Running `black .` (on `manim/`) is enough.

Author: I agree that remembering to run black is annoying. But if you remember (for example, if it is explicitly described in the uigdelines! ;) ), running the command `black manim/` is fast and easy. There's literally no input necessary from the contributor/user.

> 6. To request the ManimCommunity dev team to incorporate the changes in your

Reviewer 3: Should this be a 9?

Author: Done

   fork into the main repository, you can make a Pull Request to the Manim
   Community repo from your fork. Make sure to select `ManimCommunity/manim`
   instead of `3b1b/manim` as the `base repository` and your fork and branch as
   `head repository` - see the picture below.

   ![pull-requests-example-manim-community](./readme-assets/pull-requests.PNG)

Note: before submitting a PR please make sure that you update the docstrings of
any methods, classes, or modules that your PR modifies, if necessary.

Note: if you are contributing new features or bug fixes, it is strongly
preferred that you include tests that accompany them in your PR.  For this
purpose, you can create a new file under the `tests/` folder and make sure that

> your new tests pass before submitting your PR.

Reviewer 6: ```suggestion
your new tests pass before submitting your PR. Please check out the [Wiki Entry for Testing](https://github.com/ManimCommunity/manim/wiki/Testing) to learn more about how our tests work. 
```

Link them to the Wiki for further reading.

Note: if you are contributing documentation, you must also make sure that the
command `make doctest`, executed from the `docs/` folder terminates without

> problems.

Reviewer 6: ```suggestion
problems. Please check out the [Wiki Entry for Documentation](https://github.com/ManimCommunity/manim/wiki//Documentation-guidelines-(WIP)) to learn more about how we prefer our code documented.
```

Link them to the Wiki for further reading.

## Other guidelines

1. When submitting a PR, please make special note of whether your proposed
   changes will result in breaking changes.

> 1. When submitting a PR, make sure that your proposed changes are as general as
>    possible, and ready to be taken advantage by all of manim's users.  In
>    particular, leave out any machine-specific configurations, or any personal
>    information it may contain.
> 
> 2. If you are a maintainer, please label issues and PRs appropriately and
>    frequently.
> 
> 3. When opening a new issue, if there are old issues that are related, link
>    them in your new issue (even if the old ones are close).
> 
> 4. If you find stale or inactive issues that seem to be irrelevant, please post
>    a comment saying 'This issue should be closed', and a community developer
>    will take a look.
> 
> 5. Please do as much as possible to keep issues, PRs, and development in
>    general as tidy as possible.

Reviewer 6: ```suggestion
2. When submitting a PR, make sure that your proposed changes are as general as
   possible, and ready to be taken advantage of by all of manim's users.  In
   particular, leave out any machine-specific configurations, or any personal
   information it may contain.

3. If you are a maintainer, please label issues and PRs appropriately and
   frequently.

4. When opening a new issue, if there are old issues that are related, link
   them in your new issue (even if the old ones are closed).

5. If you find stale or inactive issues that seem to be irrelevant, please post
   a comment saying 'This issue should be closed', and a community developer
   will take a look.

6. Please do as much as possible to keep issues, PRs, and development in
   general as tidy as possible.
```

Renumbered the points; fixed a typo.

>    general as tidy as possible.

Reviewer 4: Here it would be nice to also add a list of these nice tags that @leotrs  suggested for code reviews like nitpicking, suggestion, etc.

Author: Good catch. Added link

   issue.
   
7. Please give the community a reasonable amount of time before asking again,
   or insisting on your issue.  Keep in mind that everyone is a volunteer.  If
   you wait for a reasonable amount of time and you receive no response, feel
   free to ask again.

## Contributing code

Many ways of contributing will involve writing, reading, testing, or
refactoring code.  As our repository is a Fork of [Manim by
3b1b](https://github.com/3b1b/manim), contributing in this way can be a bit
confusing.  Here is a short guide on how to do it.

1. First, make a fork of this repository.  This creates your own copy of the
   whole codebase on GitHub.

2. Clone the repository you forked using the command below (bash/zsh).

   ```sh
     git clone <your-fork-url>
     cd manim
   ```

	Note: do not `git clone` the original ManimCommunity repository.  You must
    clone your own fork.  After this step, there are three different
    repositories to keep track of: the original ManimCommunity repo, your own
    fork of it, and your local repository.

3. You have to make `git` be aware of the three repositories, by entering the

>    commands below.

Reviewer 5: Isn't it two repos?

Author: Fair

   ```sh
   git remote add upstream https://github.com/ManimCommunity/manim.git

>    git pull upstream master

Reviewer 5: **Issue**: It should be `git fetch` and not `git pull`. `git pull` will create a merge commit which is unnecessary hear and could cause problems.
```suggestion
   git fetch upstream
```

   ```

	After these commands, your local repository can keep track of your fork
    (referred to as 'origin') as well as the main ManimCommunity repository
    (referred to as 'upstream').

4. Choose the correct branch for your changes.  If you are working on a branch
   that already exists on the main repository, you can execute
   ```sh
   git checkout -b <branch-name> upstream/master
   ```
   
   If you are starting a new branch, execute
   ```sh
   git checkout -b <new_branch-name>
   ```

	You can now make any changes to your local repository, add them, and commit
   them.

5. To make your changes visible in your GitHub fork, instead of typing in `git
   push` as usual, you need to enter the command below.

   ```sh
   git push -u origin <branch-name>
   ```

   Doing so creates a new branch with the updated contents of your fork on
   GitHub (the 'origin').

6. If you want your changes to be incorporated to the main ManimCommunity
   repository, you need to make sure that there are no merge conflicts between
   the current upstream/master and the changes you are trying to make.  For

>    this purpose, please execute `git pull origin master` before you push

Reviewer 5: Isn't it `git pull upstream master`?
```suggestion
   this purpose, please execute `git pull upstream master` before you push
```
This statement mentioned in original code will just pull the master branch of their fork which is unnecessary, it should pull from master branch on manim community.
